src/crimson/os/seastore/seastore_types.h

```diff
@@ -1065,15 +1207,178 @@ class laddr_t {
   static constexpr unsigned UNIT_SIZE = 1 << UNIT_SHIFT; // 4096
   static constexpr unsigned UNIT_MASK = UNIT_SIZE - 1;
 
+  // This factory is only used in nbd driver and test cases.
+  // Cast the byte offset to valid rados object extents format.
   static laddr_t from_byte_offset(loffset_t value) {
     assert((value & UNIT_MASK) == 0);
-    return laddr_t(value >> UNIT_SHIFT);
+    // make value block aligned.
+    value >>= UNIT_SHIFT;
+    laddr_t addr;
+    // set block offset directly.
+    addr.value = value & layout::BlockOffsetSpec::MASK;
+    auto id = value >> layout::BlockOffsetSpec::length;
+    // avoid conflict with global metadata extents.
+    if (id == 0) {
+      id++;
+    }
+    // move the remaining bits to local object id. id is possible
+    // to be greater than max local object id, we allow its bits overflow
+    // to reversed hash field.
+    addr.value |= Unsigned(id)
+	<< layout::ObjectContentSpec::length;
+    // don't use LOCAL_CLONE_ID_ZERO, as it is for onode extents.
+    addr.set_local_clone_id(1);
+    return addr;
   }
 
   static constexpr laddr_t from_raw_uint(Unsigned v) {
     return laddr_t(v);
   }
 
+  // Return wheter this address belongs to global metadata(RootBlock
+  // or CollectionNode).
+  // Always ignore the upgrade bit.
+  bool is_global_address() const {
+    return (value & layout::ObjectInfoSpec::MASK) == 0;
+  }
+
+  // Return wheter this address belongs to SeastoreNodeExtent
+  // Always ignore the upgrade bit.
+  bool is_onode_extent_address() const {
+    return !is_global_address() &&
+	get_local_object_id() == LOCAL_OBJECT_ID_ZERO &&
```

Comment:
Is this simpler (see `gen_object_hint()`) if we only require local-object-id to be 0 for onodes, and non-0 for objects ?